sklearn/metrics/classification.py

```diff
@@ -1715,6 +1716,9 @@ def log_loss(y_true, y_pred, eps=1e-15, normalize=True, sample_weight=None,
     y_pred = check_array(y_pred, ensure_2d=False)
     check_consistent_length(y_pred, y_true, sample_weight)
 
+    bool_multilabel = (is_multilabel(np.asarray(y_true)) and
```

Comment:
we need to call `type_of target`. if binary use the `LabelBinarizer` and apply the normalization; if `multilabel-indicator`, compute the loss directly.

Reply:
Well, LabelBinarizer has a flag for if it received multilabel input.

Reply:
type_of_target evaluates one-hot matrix (i.e. multi-class case) to 'multilabel-indicator' therefore we can't separate multi-class and multi-label using type_of_target.

ideally, type_of_target evaluates a one-hot matrix as 'multiclass-indicator' so we can handle it differently. LabelBinarizer throws ValueError when it received 'multiclass-multioutput'.

Is it a bug in type_of_target to evaluate one-hot encoded matrix (for multi-class case) to 'multilabel-indicator'?

Reply:
multiclass is a 1D array of integer if I am not wrong and not one hot encoded.

Reply:
true, but often in multi-class problems, labels are represented as one-hot encoded matrix as well as 1D array such as below:
```
>>> iris_y
array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])
```
but yet one-hot matrix is evaluated as 'multilabel-indicator' in type_of_target I think is strange. multilabel-indicator should look like have IMHO:
```
>>> multilabel
array([[1, 1, 0],
       [0, 1, 0],
       [1, 1, 1],
       [0, 0, 1]])
```

Reply:
@jnothman LabelBinarizer raises ValueError if it's 'multiclass-multioutput' or 'continuous-multioutput' but seems to be fine with multilabel-indicator. I am little bit confused as to multilabel-indicator in sklearn is not what I think multilabel is.

Reply:
multioutput means that each column is independent. So multiclass is:

```python
In [3]: x = np.array([1, 2, 3 ,4])

In [4]: type_of_target(x)
Out[4]: 'multiclass'
```

A multiclass-multioutput will be a 2d array in which each column is a difference classification problem. Note that the metrics do not support this case for the moment.

```python
In [6]: x = np.array([[1, 2, 3 ,4], [1, 2, 3, 4]]).T

In [7]: type_of_target(x)
Out[7]: 'multiclass-multioutput'
```

multilabel is a 2d array in which each column can be dependent:

```python
In [8]: x = np.array([[1, 0, 0 ,0], [1, 1, 0, 1]]).T

In [9]: type_of_target(x)
Out[9]: 'multilabel-indicator'
```

> but yet one-hot matrix is evaluated as 'multilabel-indicator' in type_of_target I think is strange. multilabel-indicator should look like have IMHO:

The thing is that we do not support one-hot encoded multiclass as this is done in keras.
Multiclass is just the 1d array of integer.